Builder/Builder/Builder.csproj

```diff
@@ -149,4 +155,19 @@
     <Analyzer Include="..\packages\Microsoft.CodeAnalysis.Analyzers.2.6.0\analyzers\dotnet\cs\Microsoft.CodeAnalysis.CSharp.Analyzers.dll" />
   </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <PropertyGroup>
+    <PostBuildEvent>del "$(TargetDir)\Microsoft.Build.dll"
+del "$(TargetDir)\Microsoft.Build.Framework.dll"
+del "$(TargetDir)\Microsoft.Build.Tasks.Core.dll"
+del "$(TargetDir)\Microsoft.Build.Utilities.Core.dll"
```

Comment:
❓ Where are these getting copied to the output directory? I don't see a reference to them.

❗️ `PostBuildEvent` is not a reliable build action. Depending on the answer to the previous question, a different approach should be employed to ensure the files are not included in the build output directory.

Reply:
> Where are these getting copied to the output directory? I don't see a reference to them.

I described this in the PR message: "Today, the package that includes `MSBuildWorkspace` pulls the MSBuild dlls into your application."

If you're interested in more detail, you can see these package references on `Microsoft.CodeAnalysis.Workspaces.Desktop` (the assembly that contains `MSBuildWorkspace` in Roslyn <= 2.8) here: https://github.com/dotnet/roslyn/blob/dev15.7.x/src/Workspaces/Core/Desktop/Workspaces.Desktop.csproj#L24-L25.

`Microsoft.CodeAnalysis.Workspaces.MSBuild` (Roslyn >= 2.9) defines these package references properly with `ExcludeAssets="Runtime"` and `PrivateAssets="All"`: https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/MSBuild/Workspaces.MSBuild.csproj#L28-L29.

> `PostBuildEvent` is not a reliable build action.

Could you describe the scenario where that is the case? That's a pretty hyperbolic statement about something that many, many projects out there use.

Reply:
> If you're interested in more detail, you can see these package references on Microsoft.CodeAnalysis.Workspaces.Desktop (the assembly that contains MSBuildWorkspace in Roslyn <= 2.8) here: 

When referencing the package via NuGet, I originally expected the dependencies to be pulled from [this nuspec](https://github.com/dotnet/roslyn/blob/master/src/NuGet/Microsoft.CodeAnalysis.Workspaces.Common.nuspec) rather than the package references in the source project. It turns out the implementation is a bit more complex. Specifically, it turns out that **ResolveAssemblyReferences** can now find dependencies of referenced assemblies:

https://github.com/Microsoft/msbuild/blob/f72bd2eaa37476819ab7647f63d445df8055dab6/src/Tasks/Microsoft.Common.CurrentVersion.targets#L2122

MSBuild log viewer shows this task is picking up several dependencies during a simple console project build that references Microsoft.CodeAnalysis.Workspaces.Common version 2.7.0:

![image](https://user-images.githubusercontent.com/1408396/39004233-77749e80-43c2-11e8-9263-4926d19255e4.png)

> Could you describe the scenario where that is the case? That's a pretty hyperbolic statement about something that many, many projects out there use.

I actually meant the statement to be literal. Since the `PostBuildEvent` doesn't tie into up-to-date checks or conditional task execution, there are many scenarios where it will simply not execute. In past work to extend MSBuild in various ways, I ran into several cases where builds were misbehaving in edge cases for this reason, and the solution most people accepted was "just rebuild everything". When correctness of the build result depends on an action, I try to tie it more closely to the build process to eliminate the chance of an incremental build failing to produce the correct result, without getting into situations where a rebuild might be necessary. In this case, I found the following target worked well:

```diff
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <Target Name="DoNotCopyMSBuildDependencies" AfterTargets="ResolveAssemblyReferences">
+    <ItemGroup>
+      <ReferenceCopyLocalPaths Remove="@(ReferenceCopyLocalPaths)"
+                               Condition="'%(Filename)%(Extension)' == 'Microsoft.Build.dll'
+                                       OR '%(Filename)%(Extension)' == 'Microsoft.Build.Framework.dll'
+                                       OR '%(Filename)%(Extension)' == 'Microsoft.Build.Tasks.Core.dll'
+                                       OR '%(Filename)%(Extension)' == 'Microsoft.Build.Utilities.Core.dll'" />
+    </ItemGroup>
+  </Target>
 </Project>
```

Reply:
Fair enough. The key point is what I wrote in the PR description:

> The `Microsoft.Build.*` assemblies need to be deleted from your output folder.

If a `PostBuildEvent` works, great! If it's not working, your solution is a reasonable one.

Reply:
💯 to @sharwell's suggestion. We should strive to get people our of `PostBuildEvent`, it only causes pain in the long run for everyone.

Reply:
Just found that the latest version of the locator package now suggests a better way, just add `ExcludeAssets="runtime"` to your MSBuild package references :)

Reply:
@kzu: That is definitely the preferred approach and the new `Microsoft.CodeAnalysis.Workspaces.MSBuild` package does. However, this particular example uses packages.config. I'm not sure that's available to us here.